app/code/Magento/Catalog/Helper/Product/Compare.php

```diff
@@ -324,8 +324,7 @@ public function calculate($logout = false)
         $collection->addPriceData()
             ->setVisibility($this->_catalogProductVisibility->getVisibleInSiteIds());
 
-        $count = $collection->getSize();
-        $this->_catalogSession->setCatalogCompareItemsCount($count);
+        $this->_catalogSession->setCatalogCompareItemsCount(count($collection->getItems()));
```

Comment:
Provided changes introduce performance degradation on several scenarios: 
```
Simple Product 1 Comparison Add (Compare Products): 20.8% (mainline = 24 | team = 29) -> DEGRADATION
Simple Product 2 Comparison Add (Compare Products): 22.7% (mainline = 22 | team = 27) -> DEGRADATION
Configurable Product 1 Comparison Add (Compare Products): 19.2% (mainline = 26 | team = 31) -> DEGRADATION
```

Please, refactor your solution by modifying `getSelectCountSql` method and fetch only records count instead of all items info.